[istio] Add Deckhouse-flavored istioctl builds for debug container (#21160)
deckhouse/deckhouse: modules/110-istio/docs/EXAMPLES.md

### Allow from any cluster (via mTLS)

{% alert level="warning" %}

> The denying rules (if they exist) have priority over any other rules. For details, refer to the [Decision-making algorithm](#decision-making-algorithm).

Reviewer: ```suggestion
The denying rules (if they exist) have priority over any other rules. For details, refer to ["Decision-making algorithm"](#decision-making-algorithm).
```

The module-managed `Telemetry` objects for this mode disappear on the next sync; Istio restores the full `telemetry.v2` configuration.

## Debugging Istio with istioctl from the debug container

The DKP debug container includes versioned `istioctl` binaries. Use it when you need to inspect Istio configuration, run analyzers, or retrieve Envoy proxy configuration from application Pods.

> Before starting the debug container, create a dedicated ServiceAccount and grant it the permissions required by the `istioctl` commands you want to run. For example, the following manifest allows `istioctl proxy-config` commands for Pods in one application namespace:

Reviewer: ```suggestion
Before starting the debug container, create a dedicated ServiceAccount and grant it the permissions required by the `istioctl` commands you want to run. For example, the following manifest grants permissions that allow running the `istioctl proxy-config` commands for Pods in a single application namespace:
```

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: istioctl-debug
  namespace: <debug-namespace>
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: istioctl-debug
  namespace: <target-namespace>
rules:
- apiGroups: [""]
  resources:
  - pods
  verbs:
  - get
  - list
- apiGroups: [""]
  resources:
  - pods/portforward
  verbs:

>   - create

Reviewer: ```suggestion
rules:
  - apiGroups: [""]
    resources:
      - pods
    verbs:
      - get
      - list
  - apiGroups: [""]
    resources:
      - pods/portforward
    verbs:
      - create
```

---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: istioctl-debug
  namespace: <target-namespace>
subjects:
- kind: ServiceAccount
  name: istioctl-debug

>   namespace: <debug-namespace>

Reviewer: ```suggestion
  - kind: ServiceAccount
    name: istioctl-debug
    namespace: <debug-namespace>
```

roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: istioctl-debug
```

Replace `<debug-namespace>` with the namespace where the temporary debug Pod will be created, and `<target-namespace>` with the namespace of the application Pod you want to inspect. Repeat the `Role` and `RoleBinding` for every target namespace where `istioctl` must access Pods.

> This RBAC is intended for commands that address a Pod directly, for example `<pod-name>.<target-namespace>`. If you use typed resource names such as `deployment/<name>`, grant additional read access to those resource types so `istioctl` can resolve them to Pods.

Reviewer: ```suggestion
Replace `<debug-namespace>` with the namespace where the temporary debug Pod will be created, and `<target-namespace>` with the namespace of the application Pod you want to inspect. Create the Role and RoleBinding resources for every target namespace where `istioctl` must access Pods.

This RBAC manifest is intended for commands that address a Pod directly, for example, to a resource like `<pod-name>.<target-namespace>`. If you use typed resource names such as `deployment/<name>`, grant additional read access to those resource types so `istioctl` can resolve them to Pods.
```

{% alert level="warning" %}

> Creating Pods in system namespaces such as `d8-system` and using system ServiceAccounts such as `deckhouse` usually requires cluster-admin privileges. Use a dedicated ServiceAccount with the minimum required permissions instead.

Reviewer: ```suggestion
Creating Pods in system namespaces such as `d8-system` and using system ServiceAccounts such as `deckhouse` usually requires the `cluster-admin` level of privileges. Use a dedicated ServiceAccount with the minimum required permissions instead.
```

{% endalert %}

Start a temporary debug Pod with the built-in debug image:

```shell
IMG="$(d8 k -n d8-system get cm debug-container -o jsonpath='{.data.image}')"

d8 k -n <debug-namespace> run istioctl-debug \
  --rm -it \
  --restart=Never \
  --image="$IMG" \
  --overrides='{"spec":{"serviceAccountName":"istioctl-debug","automountServiceAccountToken":true}}' \
  -- bash
```

> Select the Istio minor version used by the target control plane:

Reviewer: ```suggestion
Select the minor version of Istio used by the target control plane:
```

```shell
export ISTIOCTL_VERSION=1.21
```

Available values are `1.21`, `1.25`, and `1.27`. You can also run a specific binary directly: `istioctl-1.21`, `istioctl-1.25`, or `istioctl-1.27`.

Example:

```shell
istioctl pc all <pod-name>.<target-namespace>
```

> The `istioctl pc` commands require a target Pod with an injected `istio-proxy` sidecar. If the target workload has no sidecar, Envoy admin port `15000` will not be available.

Reviewer: ```suggestion
The `istioctl pc` commands require a target Pod with an injected `istio-proxy` sidecar. If the target Pod has no sidecar, Envoy admin port `15000` will not be available.
```
